### PR TITLE
Fomo nonmonotone extension

### PR DESCRIPTION
**Pull request by d-monnet**
Add non-monotone strategy to `fomo` solver.
Modifies conditions on momentum contribution.

Related paper can be shared upon request.

### CODE REVIEW

File: src/fomo.jl [hunk 1]
```diff
@@ -341,16 +356,16 @@ function SolverCore.solve!(
...
     step_underflow = x == c # step addition underfow on every dimensions, should happen before solver.α == 0
-    ΔTk = ((oneT - βmax) * norm_∇fk^2 + βmax * mdot∇f) * λk # = dot(d,∇fk) * λk with momentum, ‖∇fk‖²λk without momentum
+    ΔTk = ((oneT - βmax) * norm_∇fk^2 + βmax * mdot∇f) * μk # = dot(d,∇fk) * μk with momentum, ‖∇fk‖²μk without momentum
     fck = obj(nlp, c)
     if fck == -Inf
       set_status!(stats, :unbounded)
       break
     end
-    ρk = (stats.objective - fck) / ΔTk
+    ρk = (max_obj_mem - fck) /(max_obj_mem - stats.objective + ΔTk)
```
Comment (tmigot) [inline src/fomo.jl:368]:
```suggestion
    ρk = (max_obj_mem - fck) / (max_obj_mem - stats.objective + ΔTk) 
```

File: src/fomo.jl [hunk 2]
```diff
@@ -369,13 +384,16 @@ function SolverCore.solve!(
         momentum .= ∇fk .* (oneT - β) .+ momentum .* β
       end
       set_objective!(stats, fck)
+      circshift!(obj_mem,1)
```
Comment (tmigot) [inline src/fomo.jl:387]:
```suggestion
      circshift!(obj_mem, 1)
```

File: src/fomo.jl [hunk 3]
```diff
@@ -429,18 +447,19 @@ function SolverCore.solve!(
 end
 
 """
-    find_beta(m, mdot∇f, norm_∇f, β, θ1, θ2)
+find_beta(m, mdot∇f, norm_∇f, μk, fk, max_obj_mem, β, θ1, θ2)
```
Comment (tmigot) [inline src/fomo.jl:450]:
```suggestion
    find_beta(m, mdot∇f, norm_∇f, μk, fk, max_obj_mem, β, θ1, θ2)
```

File: src/fomo.jl [hunk 4]
```diff
@@ -20,6 +20,9 @@ mk .= ∇f(xk) .* (1 - βmax) .+ mk .* βmax
 and βmax ∈ [0,β] chosen as to ensure d is gradient-related, i.e., the following 2 conditions are satisfied:
 (1-βmax) .* ∇f(xk) + βmax .* ∇f(xk)ᵀmk ≥ θ1 * ‖∇f(xk)‖² (1)
 ‖∇f(xk)‖ ≥ θ2 * ‖(1-βmax) *. ∇f(xk) + βmax .* mk‖       (2)
+In the nonmonotone case, (1) rewrites
+(1-βmax) .* ∇f(xk) + βmax .* ∇f(xk)ᵀmk + (fm - fk)/μk≥ θ1 * ‖∇f(xk)‖²,
+with fm the greatest objective value over the last M successful iterations, and fk = f(xk).
```
[the remark this answers is not in the document]
Comment (tmigot) [reply src/fomo.jl:25]: Ok, now I understand, it is the largest value over the last M iterations, if so, not need for the `best` indeed.

**Review by tmigot — CHANGES_REQUESTED** [review]
Thanks @d-monnet for the PR! I made a first batch of comments.
You should also add unit tests for this variant.

File: src/fomo.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (d-monnet) [reply src/fomo.jl:25]: Yes it is, though largest is a better choice of word that greatest.

[the remark this answers is not in the document]
Comment (tmigot) [reply src/fomo.jl:25]: Agreed, `largest` would be better.

File: src/fomo.jl [hunk 5]
```diff
@@ -429,18 +456,29 @@ function SolverCore.solve!(
 end
 
 """
-    find_beta(m, mdot∇f, norm_∇f, β, θ1, θ2)
+    find_beta(m, mdot∇f, norm_∇f, μk, fk, max_obj_mem, β, θ1, θ2)
 
-Compute value `βmax` that saturates the contribution of the momentum term to the gradient.
-`βmax` is computed such that the two gradient-related conditions are ensured: 
-1. (1-βmax) * ‖∇f(xk)‖² + βmax * ∇f(xk)ᵀm ≥ θ1 * ‖∇f(xk)‖²
+Compute βmax which saturates the contibution of the momentum term to the gradient.
```
Comment (tmigot) [inline src/fomo.jl:461]:
```suggestion
Compute βmax which saturates the contribution of the momentum term to the gradient.
```

**Review by tmigot — COMMENTED** [review]
@d-monnet Would you have a simple example or an illustration of the nonmonotone vs. monotone?

**Comment by d-monnet** [comment]
> @d-monnet Would you have a simple example or an illustration of the nonmonotone vs. monotone?

Not a simple example, but here are the perf profile (iterations) for `R2`, `TR`, and `fomo` with r2 and tr steps.
Profiles were obtained on 74 unconstrained problems of ADNLPProblems, run with Float64 with `atol=rtol = 1e-4` and default parameters.

FOMO(R2)_1000
FOMO(TR)_1000
R2_1000
TR_1000

File: src/fomo.jl [hunk 6]
```diff
@@ -429,18 +456,29 @@ function SolverCore.solve!(
...
 
-Compute value `βmax` that saturates the contribution of the momentum term to the gradient.
-`βmax` is computed such that the two gradient-related conditions are ensured: 
-1. (1-βmax) * ‖∇f(xk)‖² + βmax * ∇f(xk)ᵀm ≥ θ1 * ‖∇f(xk)‖²
+Compute βmax which saturates the contribution of the momentum term to the gradient.
+`βmax` is computed such that the two gradient-related conditions (first one is relaxed in the nonmonotone case) are ensured: 
+1. (1-βmax) * ‖∇f(xk)‖² + βmax * ∇f(xk)ᵀm + (max_obj_mem - fk)/μk ≥ θ1 * ‖∇f(xk)‖²
 2. ‖∇f(xk)‖ ≥ θ2 * ‖(1-βmax) * ∇f(xk) .+ βmax .* m‖
-with `m` the momentum term and `mdot∇f = ∇f(xk)ᵀm` 
+with `m` the momentum term and `mdot∇f = ∇f(xk)ᵀm`, `fk` the model at s=0, `max_obj_mem` the largest value of objective over the last M successful iterations.
```
Comment (tmigot) [inline src/fomo.jl:465]:
```suggestion
with `m` the momentum term and `mdot∇f = ∇f(xk)ᵀm`, `fk` the model at s=0, `max_obj_mem` the largest objective value over the last M successful iterations.
```

**Review by tmigot — APPROVED** [review]
Thanks @d-monnet it looks good to me! Just one last typo.

When updating your branch to the current code, in JSO, we use `git rebase` instead of `git merge`. This would avoid duplicating the changes in your PR.

**Comment by d-monnet** [comment]
> Thanks @d-monnet it looks good to me! Just one last typo.
> 
> When updating your branch to the current code, in JSO, we use `git rebase` instead of `git merge`. This would avoid duplicating the changes in your PR.

That's what I was planning to do, but I messed something up with VS Code interface...

**Comment by d-monnet** [comment]
> > > @d-monnet Would you have a simple example or an illustration of the nonmonotone vs. monotone?
> > 
> > 
> > Not a simple example, but here are the perf profile (iterations) for `R2`, `TR`, and `fomo` with r2 and tr steps. Profiles were obtained on 74 unconstrained problems of ADNLPProblems, run with Float64 with `atol=rtol = 1e-4` and default parameters.
> > Looks like we have a bug with the labels when using `label` kwargs with `profile_solvers` by the way: 2 by 2 comparison labels are wrong.
> > FOMO(R2)_1000 FOMO(TR)_1000 R2_1000 TR_1000
> 
> That's awesome! In case, you still have the result of this benchmark, maybe you can find an example of a problem that has been solved by nonmonotone that has not been solved by monotone? I think we should plan a relatively short tutorial on this 1D methods and such an example would be really great.

I'll look into it and start working on a tutorial.

**Pull request merged**